dali/kernels/imgproc/geom/remap_npp_test.cc

```diff
@@ -0,0 +1,295 @@
+#include "dali/kernels/context.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/test/cv_mat_utils.h"
+#include "dali/test/mat2tensor.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+#include "include/dali/core/tensor_shape_print.h"
+
+namespace dali::kernels::remap::test {
+
+using namespace std;  // NOLINT
+
+namespace {
+
+/**
+ * Allocate test data in managed memory (to be accessible both from CPU and GPU).
+ *
+ * Might be used to generate input data as well as remap parameters (i.e. maps).
+ *
+ * This function returns 3 entities:
+ * 1. TensorView, that wraps a buffer (3.)
+ * 2. cv::Mat, that wraps a buffer (3.)
+ * 3. A buffer, which contains test data. The caller is responsible for calling `cudaFree`
+ *    on this buffer.
+ *
+ * All three returned entities correspond to a single test case.
+ *
+ * @tparam Backend StorageBackend. Typically StorageUnified.
+ * @tparam T Type of the test data.
+ * @tparam ndims Number of dimensions in the generated data.
+ * @tparam Generator Function, that suffices signature `T g();` and returns a random number.
+ * @param shape Shape of the test data.
+ * @param stream The stream, on which the copy will be conducted.
+ * @return
+ */
+template<typename Backend = StorageUnified, typename T, int ndims, typename Generator>
+tuple<TensorView<Backend, T, ndims>, cv::Mat, T *>
+alloc_test_data(TensorShape<ndims> shape, Generator g, cudaStream_t stream = 0) {
+  using U = remove_const_t<T>;
+  vector<U> source_data(volume(shape));
+  generate(source_data.begin(), source_data.end(), g);
+  remove_const_t<U> *ret;
+  CUDA_CALL(cudaMallocManaged(&ret, volume(shape) * sizeof(T)));
+  CUDA_CALL(cudaMemcpyAsync(ret, source_data.data(), volume(shape) * sizeof(T), cudaMemcpyDefault,
+                            stream));
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  auto nchannels = shape.sample_dim() < 3 ? 1 : shape[2];
+  cv::Mat cv_mat(shape[0], shape[1], CV_MAKETYPE(cv::DataDepth<U>::value, nchannels), ret);
+  return make_tuple(TensorView<Backend, T, ndims>(ret, shape), cv_mat, ret);
+}
+
+
+/**
+ * Function that counts outlying pixels. This function is used for verifying the remap result.
+ *
+ * Outlying pixels for given two images are the pixels, that do not match the corresponding one.
+ * In other words, these are the pixels that are not (0,0,0) in the difference of the images.
+ *
+ * @tparam T Type of the input data.
+ * @param diff The difference of two images (img1 - img2).
+ * @return Number of outlying pixels.
+ */
+template<typename T>
+size_t count_outlying_pixels(const cv::Mat &diff) {
+  const auto &p = reinterpret_cast<const T *>(diff.data);
+  int ret = 0;
+  for (int i = 0; i < diff.rows * diff.cols * diff.channels(); i += diff.channels()) {
+    // Effectively compare to 0, but secure from float precision.
+    auto res = std::accumulate(p, p + diff.channels(), 0.f);
+    if (-1e-7 > res || res > 1e-7) ret++;
+  }
+  return ret;
+}
```

Comment:
This implementation is flawed - OpenCV matrices can have row padding.
```suggestion
size_t count_outlying_pixels(const cv::Mat &diff) {
  cv::Mat tmp = cv::abs(diff);
  cv::threshold(tmp, tmp,1e-7, 1.0, cv.THRESH_BINARY);
  return cv::countNonZero(tmp);
}
```

Reply:
Unfortunately, `cv::countNonZero` works on 1-channel images, while here I need to operate on both 1-channel and 3-channel. As I wrote in the comment above, the cv::Mats in this file wrap the contiguous memory, so maybe this function isn't bad after all?

Reply:
In that case a big fat comment is in order - and perhaps also an assert
```
assert(mat.ptr<T>(1) - mat.ptr<T>(0) == mat.cols);
```
or something like that.

Reply:
Added comment and assert